placement/placement.go

```diff
@@ -0,0 +1,667 @@
+		return nil
+	}))
+}
+
+func (sp storagePlacement) DecommissionCluster(dbName, cName string) error {
+	return sp.mgr.Change(wrapFn(func(p *schema.Placement, changes *schema.PlacementChanges) error {
+		db, dbChanges := sp.findDatabase(p, changes, dbName)
+		if db == nil {
+			return errDatabaseNotFound
+		}
+
+		// If this is an existing cluster, add to the list of decomissions if it's not already there
+		if _, existing := db.Clusters[cName]; existing {
+			if dbChanges.Decomms == nil {
+				dbChanges.Decomms = make(map[string]*schema.ClusterDecommission)
+			}
+
+			sp.log.Infof("decommissioning cluster %s in database %s", cName, dbName)
+			dbChanges.Decomms[cName] = &schema.ClusterDecommission{
+				ClusterName: cName,
+			}
+			return nil
+		}
+
+		// If this is a pending join, delete from the join list
+		if _, joining := dbChanges.Joins[cName]; joining {
+			sp.log.Infof("removing pending cluster %s from database %s", cName, dbName)
+			delete(dbChanges.Joins, cName)
+			return nil
+		}
+		return errClusterNotFound
+	}))
+}
+
+func (sp storagePlacement) GetPendingChanges() (int, *schema.Placement, *schema.PlacementChanges, error) {
+	vers, protoConfig, protoChanges, err := sp.mgr.GetPendingChanges()
+
+	var p *schema.Placement
+	if protoConfig != nil {
+		p = protoConfig.(*schema.Placement)
+	}
+
+	var changes *schema.PlacementChanges
+	if protoChanges != nil {
+		changes = protoChanges.(*schema.PlacementChanges)
+	}
+
+	return vers, p, changes, err
+}
+
+func (sp storagePlacement) CommitChanges(version int, opts CommitOptions) error {
+	return sp.mgr.Commit(version, wrapFn(func(p *schema.Placement, changes *schema.PlacementChanges) error {
+		sp.log.Infof("committing changes at version %d", version)
+
+		// Pull all existing databases, and order them by max retention period.  This will be used
+		// to determine if any new databases are pulling traffic from an existing database
+		existingDatabases := make([]*schema.Database, 0, len(p.Databases))
+		for _, existing := range p.Databases {
+			existingDatabases = append(existingDatabases, existing)
+		}
+		sort.Sort(databasesByRetention(existingDatabases))
+
+		// Add new databases
+		for _, add := range changes.DatabaseAdds {
+			if err := sp.commitNewDatabase(p, add.Database, existingDatabases, opts); err != nil {
+				return err
+			}
+		}
+
+		// Process database changes, including those from new databases
+		for name, change := range changes.DatabaseChanges {
+			if err := sp.commitDatabaseChanges(p, p.Databases[name], change, opts); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}))
+}
+
+func (sp storagePlacement) commitNewDatabase(
+	p *schema.Placement,
+	db *schema.Database,
+	existingDatabases []*schema.Database,
+	opts CommitOptions) error {
+	sp.log.Infof("Creating initial database configuration for %s", db.Name)
+
+	// Check to see if there are any existing databases that cover this retention
+	// period - if so then we need to do a staged cutover to that database
```

Comment:
This is for the case where we are doing things like replacing an entire historical cluster (in this case cluster(s)) with new historical cluster(s)?  Is there any reason we wouldn't just model this as a set of cluster decommissions and joins for an existing database?